### DISCUSSION TITLE
A problem about the choice of the base learner for the PU learning

### DISCUSSION THREAD

**Discussion by XueJiaAIMR** [pulearn/pulearn#30] [Q&A]
Hello:

The reference of PU bagging is "A bagging SVM to learn from positive and unlabeled examples", but the code is not the SVM, But is the Random forest. Does that mean any classifier learner (SVC, Decision tree, Random Forest, etc) can be applied in this bagging method? and so are the Standard classifier and two-step methods ? 

Looking forward to your reply.

sincerely yours

Xue Jia

**Comment by shaypal5** [comment]
Hey Xue Jia,

I apologize for my belated response.

Regarding your question, you must differentiate the ensemble-building algorithm from the learner used.

The PU bagging method uses the bagging method to build an ensemble classifier, which can be used with any base learner `scikit-learn` provides.

Now, a few notes:
1. It doesn't mean it *should* be used with any learner. Bagging (bootstrap aggregation) methods were designed to augment high-variance, low-bias base learners, like decision trees ([see this excellent post](https://machinelearningmastery.com/bagging-and-random-forest-ensemble-algorithms-for-machine-learning/)).
2. You might be confusing bagging generally with random forest specifically. Random forest is a bagging learning method that adds an additional constraint on base learners: Each gets only a subset of features to learn on (and not only a subset of samples, like in bagging generally). This turns out to significantly reduce correlation between the individual base learners and thus improve the ensemble model results.
3. It is true that if you choose a decision tree as your base learner for the PU bagging ensemble you get a tree-based ensemble, but this is not a random forest classifier.

Regarding the two-step method, if I remember correctly, any classifier that outputs probabilities can be used as the base learner there.

Cheers,
Shay
